### ISSUE TITLE
General discussion

### ISSUE THREAD

**Issue by ghost** [DemoProductions/shmup#38]

**Comment by ghost** [comment]
More work on this project tomorrow. Going to get some extra sleep. Lets discuss what kind of game we want to make more later

**Comment by flip40** [comment]
Sounds good, I'll leave the discussion questions to start with you. Not sure what to discuss off the top of my head, though I have some ideas on mechanics we still need (probably) to add.

**Comment by ghost** [comment]
More work on this tomorrow, did not get a chance to see your pull requests today. Will think more about the kind of shmup this can be

**Comment by ghost** [comment]
A halt on development soon to just focus on discussing game ideas?

**Comment by flip40** [comment]
Yeah, we need to decide on more of the structure of the game. Should probably also see if Tyler and Jamahl are still around and want to input ideas..

**Comment by flip40** [comment]
So some of my thoughts...
i feel like a "rogue-lite" shmup would be fun to make, at least for gameplay. I was thinking the other day that taking a lot of the mechanics from FTL, a rogue-lite "top down space simulator" (if they say so..), and applying them to a shmup.

The idea would be to have a number of levels that each have a mini-boss of some sort at the end, with a final boss at the very end of these ~8 or so levels. After each mini boss (and possibly as random drops throughout the level as well) you could spend points gained throughout the level on upgrading your ship. Death would mean restarting, however during each run you would be able to unlock more content overall, such as new ships to play as (which should all have unique play styles) and such.

What ideas do you guys have?

**Comment by ghost** [comment]
> So some of my thoughts...
> i feel like a "rogue-lite" shmup would be fun to make, at least for gameplay. I was thinking the other day that taking a lot of the mechanics from FTL, a rogue-lite "top down space simulator" (if they say so..), and applying them to a shmup.
> 
> The idea would be to have a number of levels that each have a mini-boss of some sort at the end, with a final boss at the very end of these ~8 or so levels. After each mini boss (and possibly as random drops throughout the level as well) you could spend points gained throughout the level on upgrading your ship. Death would mean restarting, however during each run you would be able to unlock more content overall, such as new ships to play as (which should all have unique play styles) and such.
> 
> What ideas do you guys have?

We could also use the mechanic of recruiting new crew members. These crew members can be bought from shops or they decide to join at some point during levels (maybe some of these can be level specific) depending on the story. 

Possibilities on different types of crew members:
- "Basic": (not sure what the general name is to call this) member that tends to stay in the middle-player-end of the playing field or main camera and shoots enemies that are closest to it. Applies minor damage and is limited to using basic bullets.
- Snipers: Stays farthest basic and shoots enemies from afar. Powerful shots with a long delay between shots. Tends to shoot enemy snipers first or enemies that are "farther back."
- Tanks: Missles, higher damage, longer delay (not as long as sniper shots). High health. Stays towards middle-player-end.
- Medic/Engineer: Fixes ships somehow

And more to come if we go with the crew mechanic

Enemies drop a random set of money, depending on what type of enemy they are. Ex: Basic enemies drop 5 and 10, and More advanced enemies drop 50 and 100. 

Besides dropping money, certain enemies may drop weapons (randomly?) which the player or the player's crew can use.

Money gained from levels can be used to buy things in the shop

Categories for things that can be bought
- Spaceship parts
- Weapons
- Bombs?
- Bullets? (This may complicate things)
- Crew

Other questions:
If we go with the crew mechanic, does the game end when the crew dies or the player dies?

**Comment by ghost** [comment]
Story idea:

Player is some type of commander of some fleet during a war. At this time, humans have colonized planets in the scope of many galaxies. Both sides are fighting for some type of resource, which may be left ambiguous to the people fighting on both sides. This resource can be a single special object, a once abundant resource, or something else. As the story progresses, players and the enemies they were fighting have morality questions, and may end up fighting a single "greater" enemy that is sort of manipulating this whole thing.

**Comment by ghost** [comment]
Story idea:

Player is the leader of some type of unknown rebel group, some type of opposition against some type of corrupt government during some time in the future. Humans have also colonized many planets in the scope of many galaxies in this story. The player battles while recruiting new members from planets that the player "saved." Eventually the player reaches the "core of evil" and has some type of final epic battle there.

**Comment by ghost** [comment]
Story idea:

Player is a commander for some type of corrupt government (the player does not know). As the player progresses the player starts to question their morality and eventually turns against the government that the player works for. Player battles and reaches the main source of problems but before landing the final strike the player dies (maybe something like chips implanted in the player's body that someone can push some red button to activate). Player now controls one of the player's crew and does the final battle.

**Comment by ghost** [comment]
Story idea:

Player has always wanted to fight against the aliens that have constantly raided human planets in some distant future. But the player lacks the power needed to operate the ships. Maybe the player needs a certain resource that only certain families of higher pedigree had. One day his planet is attacked by these aliens and some person saves the player, and the player somehow ends up with this resource to operate ships. He operates spaceships naturally from all the years of studying battles. And goes on to fight 7 major battles until finally reaching the Queen or King alien.

**Comment by ghost** [comment]
Story idea:

Player is average student in the present day. One day as the player is standing on their favorite rooftop admiring life, a paper airplane hits the player. The player finds that the paper airplane is a letter and the player is somehow in this other world now. In this world, everyone is some type of orb or ship thing and people keep trying to take this thing that is inside his orb. Some Gandalf figure shows the player the way and explains some of the reason for why people want what is inside his orb. The player does a series of battles with the help of experienced crew members and eventually reaches the last battle.

**Comment by ghost** [comment]
Story idea:

Player is an average student that meets quack scientist that believes he has invented time travel. Turns out the quack scientist really did invent time travel and the player is somehow in the year 5xxx. Player spawns in middle of battlefield and mayhem ensues.

**Comment by ghost** [comment]
Story idea:

Player is a former war hero on his deathbed reminiscing on past battles. Player goes through the war hero's 8 major battles.

**Comment by ghost** [comment]
Story idea:

Player is hunting mushrooms in some forest and accidentally eats a magic mushroom. Mayhem ensues in player's head.

**Comment by ghost** [comment]
The progression of my ideas haha

**Comment by flip40** [comment]
For clarification is the "crew" idea about crew on your ship, or as it seems, other ships that would fly with you? The mechanic seems interesting but we need to determine a balance for it. I would say the player's death would end the game. The other ships should probably generally be weaker than yours, or if they do end up being very good, very expensive so you cannot have many at a time (I would think). Would have to work that out.
Either way, I think we need to be sure to focus on making sure the player has reason to play the game. If the AI are too good, you could theoretically have the AI start doing everything for you, which might be cool if you had to work really hard to get there, but if that is too easy to attain then the game might become boring for being too easy.

I haven't read all the story ideas yet, I'll give them a closer look when I get home. I did quickly read this one and kind of like it for its simplicity, yet still having some meaning and giving a story / narrative to the gameplay "Player is a former war hero on his deathbed reminiscing on past battles. Player goes through the war hero's 8 major battles."

A thought on the story stuff, if we are relying on the mechanic of the game being a RogueLite we probably need to be sure the story fits well in terms of constant restarting. I would argue if we focus too much on storytelling throughout the game, having to restart might be more painful than starting the current level over again. On the other hand, we could perhaps find a way to integrate the mechanic of restarting on death to somehow fit in with the theme of the story (say for the reminiscing one, you have to remember it just right, and if you die, well "no no, that wasn't quite what happened" or something).

Also would have to consider ease of restarting. It is a definite fact that the ability to hit the "retry" button and start over as fast as physically possible is a great way to keep people going on the "just one more try" train.

Random last thought: drones or turrets as a form of weaponry. I would still like to have different ships you could use to allow different gameplay, so focusing on a different branch of combat or something (missile focused, laser focused, defensively focused, drone focused, whatever).

**Comment by ghost** [comment]
> For clarification is the "crew" idea about crew on your ship, or as it seems, other ships that would fly with you? The mechanic seems interesting but we need to determine a balance for it. I would say the player's death would end the game. The other ships should probably generally be weaker than yours, or if they do end up being very good, very expensive so you cannot have many at a time (I would think). Would have to work that out.

Yes, the "crew" would be members of a certain team that fly with you. If we go with this idea, maybe there would be a limit to how many members you can have, I'd say about 3. Player death ending the game would make sense if the User is actually embodying the player and not choosing a certain member to control (sort of like in Fire Emblem). Player death ending the game would make lots of sense for the idea of the former war hero. 

> I haven't read all the story ideas yet, I'll give them a closer look when I get home. I did quickly read this one and kind of like it for its simplicity, yet still having some meaning and giving a story / narrative to the gameplay "Player is a former war hero on his deathbed reminiscing on past battles. Player goes through the war hero's 8 major battles."

Yeah a simple story, possible with some vague and generic elements, will work well. Point is for easier immersion in the game (and Player's can sort of self insert their own stories into the game for a better experience). I like the former war hero idea too.

> A thought on the story stuff, if we are relying on the mechanic of the game being a RogueLite we probably need to be sure the story fits well in terms of constant restarting. I would argue if we focus too much on storytelling throughout the game, having to restart might be more painful than starting the current level over again. On the other hand, we could perhaps find a way to integrate the mechanic of restarting on death to somehow fit in with the theme of the story (say for the reminiscing one, you have to remember it just right, and if you die, well "no no, that wasn't quite what happened" or something).

Lets not focus too much on story telling throughout the game (just something for the Player to immerse themselves better in like mentioned above. Maybe strive for a generic story with one or two unique elements. As for constant restarting, we can go with time travel or simply the ability to "return by death" (return by death would be an idea taken by a currently airing anime Re:Zero though). I like the idea of reminiscing and saying "no that wasn't quite what happened." Much smoother and doesn't break walls as much as when the level just somehow restarts.

> Also would have to consider ease of restarting. It is a definite fact that the ability to hit the "retry" button and start over as fast as physically possible is a great way to keep people going on the "just one more try" train.

Not sure what you mean by this actually haha. Do you mean that an accessible retry button would help Players who like to completely 100% finish levels? In that case, maybe can do sort of a separate game mode where the level auto restarts if the Player missed something? Sort of like Sudden Death mode or Perfect mode in osu!.

> Random last thought: drones or turrets as a form of weaponry. I would still like to have different ships you could use to allow different gameplay, so focusing on a different branch of combat or something (missile focused, laser focused, defensively focused, drone focused, whatever).

Ah yes, different ships for different styles of gameplay. A nice feature to have and a reason why I think games like Bastion are so fun and re-playable.

Edit: probably confused about Rogue-like games (have not played too many of them). What exactly do you mean by that in this case?

**Comment by flip40** [comment]
+1 for FE reference. Don't think switching characters generally would work well with the aesthetic reason most people play shmups though. But who knows, on the other hand I might be far too stuck in the view of what I do see games do instead of trying something different.

Looks like we generally agree and were thinking the same thing for the story stuff. I'll take all the ideas and store them on the wiki page or something and we can later narrow down the specifics. (sidenote, I'll also take most of the stuff we discuss here and put it there as well in more bullet pointy form to keep track of the key game ideas).

About the retry button, it is more about the time to restart when a player loses. If you lose, then are forced to watch a 30 second intro before being able to play the level again (for example), then that is not so good. When losing, pressing retry should instantly start the level over as rapidly as possible. I would say there are some exceptions to this of course. A few seconds of intro animation is more of a plus, despite the time it might take. The idea there being that the animation would somehow "launch" the level, as opposed to "hey, you spawned in along with everything else now go".

Rouge-**Like** games feature complete loss of progress and restarting upon death (named after a very old game called Rogue, which did this). Rouge-**Lite** games feature loss of "current" progress and restarting upon death, however it usually features meta progression (FTL is a good example, where you always restart when you die, however during play you could unlock new ships to start runs with). There are a variety of ways Rogue-Lite games might implement the meta progression. Some are more direct, like your player constantly improving so as to make more progress each time (see Rogue Legacy) where others are more indirect (like FTL). This game would be Rogue-**Lite**, due to unlocking ships and such to enable different play styles. The reason Rogue-Lite came into existence is due to the realization that most players don't find complete loss of progression fun, so having a form of meta progression allows progression to still happen in a game whose core features progress loss.

**Comment by ghost** [comment]
Switching characters would definitely be tricky and might make things too complicated. Hard to imagine switching Player's in a fast paced game, would be a high learning curve. 

Ok sounds good, I can help with the wiki too.

Retry button: Yes a way to skip animations on death or intro would be nice

In our game, say the Player is on the 4th level and died there. Would they restart on the 1st level with "status" intact? Or would they restart on the 4th level with "status" intact? Would the player have only one life? Or would they have multiple tries throughout the game, say 3? What sort of ships would you like for the Player to be able to have? 

One thing about Rogue-Lite games. Would this eventually leave the game too easy? Lets say that each run through, the player keeps the amount of coins they collected. Then they can buy the better items by farming coins at lower levels. Or would our Rogue-Lite game solely unlock the new types of ships that Players can use? I like the solely unlock new types of ships idea better, I think that was what you meant in your previous posts.

A starting possibility:

Story: Player is a former war hero on his deathbed reminiscing on past battles. Player goes through the war hero's 8 major battles.

Restart at level 1 upon three deaths. Unlock new types of ships at completion or when reached levels 3, 5, 7, 8.

Types of ships that the Player can use:
1. (Default). Balanced in speed, health, and defense. Turret bullets
2. Slow in speed, high health, high defense. Missles
3. Fast in speed, low health, low defense. Auto locking bullets (moves by itself towards enemies)
4. 

Types of enemies:
- Soldier. Move in a straight line (may curve outwards too to leave the map). Slow turret bullets that move linearly
- Tank: Missles
- Sniper
- placehold
- placehold
- Bosses

Weapon upgrades:
- Turret. 1) fire single bullets 2) double, faster refirerate and wider range 3) faster refire rate, wider range and lots of dmg
- Missle
-  Auto

Enemy drops:
- Coins
- placehold
- placehold

Item shop:
- Weapon upgrades
- Shields or barriers?
- Bombs?

Level 1: Former war hero's "breakout" battle. Comrades all perished on battlefield and left by himself.

Level 2: Has command of small fleet. 

Level 4: Has trust of upper military but sent to battle with low chance of survival. Meant as a diversion for a surprise attack from another fleet. (Survive for 5 mins or defeat boss).

Level 8:

**Comment by flip40** [comment]
> In our game, say the Player is on the 4th level and died there. Would they restart on the 1st level with "status" intact? Or would they restart on the 4th level with "status" intact?

Player would start on the 1st level, metagame progress intact. We could also make sure to have different difficulty levels perhaps to aid in not having a game too difficult for some people to complete, but still challenging on normal or hard difficulties. Of course, if FTL is any indication, 'easy' might still be quite difficult.

> One thing about Rogue-Lite games. Would this eventually leave the game too easy?

It requires different design considerations, but "sort of". I mentioned Rogue Legacy as an example of a game that did this, you can check it out. It relied on have 4 zones with their own bosses, each zone harder than the last. You outright needed a stronger character via the meta-leveling to actually defeat the harder zones (you also did not have to go back to earlier zones for the most part due to how the randomly generated map was laid out).

> Or would our Rogue-Lite game solely unlock the new types of ships that Players can use? I like the solely unlock new types of ships idea better, I think that was what you meant in your previous posts.

Yep, I think we agree on this point. We could add other unlocks than just ships perhaps, but the plan would be to not directly use meta progress to directly increase player strength.

**Comment by flip40** [comment]
My thoughts on integrating the story with gameplay and mechanics...

One thing we have in the story is a mention of different factions. Pulling some thoughts from FTL again, the main way of getting ships involved random in game events with different races. These would each start little quests to do within the game that if completed would unlock the race's ship for use in future runs (alternate versions of the ship were unlocked by completing 2 out of 3 ship specific achievements / challenges). What we might be able to do with our game in a similar way is to allow some random branching and choice of what your next level will be on your way to the final boss (which would always be the same so we don't have to worry about many endings). These levels could have some kind of unique objective that must be completed to continue the chain (otherwise defaulting to regular missions), and perhaps have a unique boss. Finishing a chain unlocks the ship for that faction to use in future runs.

Another thought its the mention of 'ethenium'. What ways could we integrate this resource as a game mechanic? One thought that came to mind was that it could be what we collect from killing enemies instead of score / points or whatever that we use to upgrade the ship or buy items between levels.

Both of these ideas might require a few changes and shaping the story you currently have set at the moment, but that isn't a bad thing. For example ethenium perhaps instead of being unknown could be known, but extremely rare and valuable (and still a great resource for weaponry and such). Kronos could have found a large supply of it and used it to improve the army's technology and that is how he was able to take over the other faction and become a threat. This would also conveniently explain why we get drops of this resource from destroying enemy ships, and as a nice mechanic for how we improve our ship to combat theirs.

**Comment by ghost** [comment]
Good idea for unlocking ships through the bonus or side quests in random game events. 

Believe we talked about Ethenium on Discord. Ethenium could be a rare resource that the enemy has a stockpile of and that they use in their ships, so enemy ships drop ethenium when they explode. Or could have Ethenium as something that is extremely rare but the bosses use them for something. The shop could also be purely point based, points increase when killing enemy ships. So would keep the "current" points and the running total. Not sure how we could incorporate that into the current story though. 

Kronos having a large supply of Ethenium is a good idea. Player collects ethenium and uses it to buy stuff. I like that idea the most so far

**Comment by ghost** [comment]
An update. I've noticed the comments and pull requests but haven't made the time to look at them. I won't be able to contribute regularly for now. Can go ahead and merge changes if I take too long to get back (though it defeats the purpose of code review and working in a team).

On a side note, I recently helped make some poke bowls for Kabam about a week ago. @flip40 Did you get to try that? And @stannardtyler Those look like great sprites so far.

**Comment by flip40** [comment]
No worries, everyone gets busy :P I'll leave my PR up as long as it is reasonable to do so. At the moment I am just moving in sprites, and the Boss AI already runs fine on the old code (the PR is just an attempt to vastly improve readability as well as make it easy to define new boss AI states) so I shouldn't need to merge it in until we do more boss AI at the latest.

Also I am pretty sure I did get to try some of the poke bowls, if I remember the correct ones :P I believe I had one with teriyaki eel and fried shrimp, if that sounds right. They were great :D
